File: taurus-protect-sdk-go/pkg/protect/mapper/change.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// ChangeFromDTO converts an OpenAPI Change to a domain Change.
func ChangeFromDTO(dto *openapi.TgvalidatordChange) *model.Change {
	if dto == nil {
		return nil
	}

	change := &model.Change{
		ID:                safeString(dto.Id),
		TenantID:          safeString(dto.TenantId),
		CreatorID:         safeString(dto.CreatorId),
		CreatorExternalID: safeString(dto.CreatorExternalId),
		Action:            safeString(dto.Action),
		EntityID:          safeString(dto.EntityId),
		EntityUUID:        safeString(dto.EntityUUID),
		Entity:            safeString(dto.Entity),
		Comment:           safeString(dto.Comment),
	}

	// Copy changes map
	if dto.Changes != nil {
		change.Changes = make(map[string]string)
		for k, v := range *dto.Changes {
			change.Changes[k] = v
		}
	}

	// Convert creation date
	if dto.CreationDate != nil {
		change.CreationDate = *dto.CreationDate
	}

	return change
}

// ChangesFromDTO converts a slice of OpenAPI Changes to domain Changes.
func ChangesFromDTO(dtos []openapi.TgvalidatordChange) []*model.Change {
	if dtos == nil {
		return nil
	}
	changes := make([]*model.Change, len(dtos))
	for i := range dtos {
		changes[i] = ChangeFromDTO(&dtos[i])
	}
	return changes
}

// CursorPaginationFromDTO converts an OpenAPI ResponseCursor to a domain CursorPagination.
func CursorPaginationFromDTO(dto *openapi.TgvalidatordResponseCursor) *model.CursorPagination {
	if dto == nil {
		return nil
	}

	return &model.CursorPagination{
		CurrentPage: safeString(dto.CurrentPage),
		HasPrevious: safeBool(dto.HasPrevious),
		HasNext:     safeBool(dto.HasNext),
	}
}

// CreateChangeResultFromDTO converts an OpenAPI CreateChangeResult to a domain CreateChangeResult.
func CreateChangeResultFromDTO(dto *openapi.TgvalidatordCreateChangeResult) *model.CreateChangeResult {
	if dto == nil {
		return nil
	}

	return &model.CreateChangeResult{
		ID: safeString(dto.Id),
	}
}

File: taurus-protect-sdk-typescript/src/mappers/change.ts
/**
 * Change mapper functions for converting OpenAPI DTOs to domain models.
 */

import type { Change, CreateChangeRequest } from '../models/change';
import { safeDate, safeInt, safeMap, safeString } from './base';

/**
 * Maps a change DTO to a Change domain model.
 */
export function changeFromDto(dto: unknown): Change | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;

  return {
    id: safeString(d.id),
    tenantId: safeInt(d.tenantId ?? d.tenant_id),
    creatorId: safeString(d.creatorId ?? d.creator_id),
    creatorExternalId: safeString(d.creatorExternalId ?? d.creator_external_id),
    action: safeString(d.action),
    entity: safeString(d.entity),
    entityId: safeString(d.entityId ?? d.entity_id),
    entityUUID: safeString(d.entityUUID ?? d.entity_uuid),
    changes: d.changes as Record<string, string> | undefined,
    comment: safeString(d.comment),
    // Map creationDate to createdAt (as in Java mapper)
    createdAt: safeDate(d.creationDate ?? d.creation_date ?? d.createdAt ?? d.created_at),
  };
}

/**
 * Maps an array of change DTOs to Change domain models.
 */
export function changesFromDto(dtos: unknown[] | null | undefined): Change[] {
  return safeMap(dtos, changeFromDto);
}

/**
 * Converts a CreateChangeRequest to an OpenAPI DTO.
 * Maps SDK 'comment' to OpenAPI 'changeComment' (matching Java ChangeMapper).
 */
export function createChangeRequestToDto(request: CreateChangeRequest): Record<string, unknown> {
  return {
    action: request.action,
    entity: request.entity,
    entityId: request.entityId,
    changes: request.changes,
    changeComment: request.comment,
  };
}
